### PR TITLE
feat(instrumentation-nats): add instrumentation nats package

### PR DESCRIPTION
**Pull request by giacomoquinalia**
## Which problem is this PR solving?

Initial instrumentation for NATS.

## Short description of the changes

This is a continuation of the initiative started in this PR: #667. The implementation follows the architectural considerations and feedback discussed in that PR, especially the insights shared by NATS maintainers (referencing [this maintainer comment](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/667#issuecomment-925184584)).

This implementation follows the patterns and structures used in existing messaging instrumentations like kafkajs and amqplib.

## Testing

The implementation has been tested locally and is functioning as expected.
- [x] Created unit tests
- [x] Verified the implementation by running a local script using the instrumentation-nats package and visualizing the traces in Grafana

## Feedback

While the implementation is stable in my test environment, I would appreciate validation from the maintainers and the community:

1. Are there specific NATS methods or edge cases currently missing that should be prioritized for this initial release?
2. Are the concerns raised in #667 sufficiently addressed here?

### CODE REVIEW

File: packages/instrumentation-nats/README.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+## Semantic Conventions
+
+This package uses `@opentelemetry/semantic-conventions` version `1.30+`, which implements Semantic Convention [Version 1.30.0](https://github.com/open-telemetry/semantic-conventions/blob/v1.30.0/docs/README.md)
+
+### Spans Emitted
+
+| NATS Operation | Span Kind | Span Name            | Operation Type | Operation Name |
+| -------------- | --------- | -------------------- | -------------- | -------------- |
+| `publish`      | Producer  | `send <subject>`     | `publish`      | `publish`      |
+| `request`      | Producer  | `send <subject>`     | `publish`      | `request`      |
```
Comment (aricart) [inline packages/instrumentation-nats/README.md]: this would have the reply subject, but really it is a pub operation that has a reply

**Comment by giacomoquinalia** [comment]
A better approach for the instrumentation was suggested by a NATS maintainer [in this issue](https://github.com/nats-io/nats.js/issues/360) and I'll update the implementation of this PR once the native support is available in the core library.

**Comment by raphael-theriault-swi** [comment]
Would you mind converting this to a draft in the meantime or closing and reopening ? This would help in making it easier to sort through GH notifications for approvers :)

File: packages/instrumentation-nats/src/instrumentation.ts [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+  }
+
+  private _unpatch(
+    moduleExports: typeof nats,
+    _moduleVersion: string | undefined
+  ): void {
+    if (isWrapped(moduleExports.connect)) {
+      this._unwrap(moduleExports, 'connect');
+    }
+    natsHeaders = undefined;
```
Comment (jrehwaldt) [inline packages/instrumentation-nats/src/instrumentation.ts]: Store and unstore `headers` for later use.
